### PR TITLE
feat: Properly handle critical alerts

### PR DESCRIPTION
**Pull request by zn3x**
Currently, there is no check on alerts received from recvmsg, even those that require to close session as per the rfc.
This commit adds alerts case handling when session termination is necessary.

### CODE REVIEW

File: src/ktls_stream.rs [hunk 1]
```diff
@@ -142,7 +142,33 @@ where
...
+                                        if alert.len() < 2 {
+                                            panic!("ktls sent an alert with invalid size");
+                                        }
+
+                                        // alert layout: [level, description]
+                                        // if we get a close_notify() or an alert with fatal level
+                                        // we should close session
+                                        if alert[1] == 0
+                                            || alert[0] == 2 {
+                                            _ = crate::ffi::send_close_notify(this.inner.as_raw_fd());
```
Comment (fasterthanlime) [inline src/ktls_stream.rs:161]: If that's us receiving a close_notify, do we really want to be sending a close_notify back?

File: src/ktls_stream.rs [hunk 2]
```diff
@@ -142,7 +142,33 @@ where
...
+                                            unsafe { libc::close(fd) };
```
Comment (fasterthanlime) [inline src/ktls_stream.rs:162]: Oh I guess we're also just closing it here.

**Comment by fasterthanlime** [comment]
Are you able to write an integration test to cover this change? (I know, I know, you're replacing a "panic!", but the panic is loud and explicit)

**Comment by zn3x** [comment]
Hello, I decided to add needed test case to already existing test so that I don't need to duplicate whole thing.

About errors, how do you think about handling them, it seems according to [this](https://github.com/facebookincubator/fizz/blob/fff6d9d49d3c554ab66b58822d1e1fe93e8d80f2/fizz/experimental/ktls/AsyncKTLSSocket.cpp#L148) alerts may return with unexpected length, would it be enough if I replace the panic with `std::io::Error`?

**Comment by fasterthanlime** [comment]
(I'll fix the compile errors, it involves a ktls-recvmsg release)

**Comment by zn3x** [comment]
Aside from minor fix I added, this looks better than my corny solution.

File: src/ktls_stream.rs [hunk 3]
```diff
@@ -300,6 +300,9 @@ where
             if let Err(e) = crate::ffi::send_close_notify(this.inner.as_raw_fd()) {
                 return Err(e).into();
             }
+            if *this.read_closed {
+                unsafe { libc::close(this.inner.as_raw_fd()) };
```
Comment (fasterthanlime) [inline src/ktls_stream.rs:304]: unless I'm confused (it's 4AM), this is what `this.inner.poll_shutdown` ends up doing, so I don't think it's necessary

**Comment by fasterthanlime** [comment]
@zn3x v4.0.0 is released with those changes: https://crates.io/crates/ktls/4.0.0
